File: baekjoon/1166/1166.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static double solve(int N, int L, int W, int H) {
        double low = 0;
        double high = Math.max(L, Math.max(W, H));

        while (low < high) {
            double mid = (low + high) / 2;
            long count = (long)(L/mid) * (long)(W/mid) * (long)(H/mid);

            if (count >= N) {
                if (low == mid) break;
                low = mid;
            }
            else {
                if (high == mid) break;
                high = mid;
            }
        }

        return high;
    }

    public static void main(String[] args) {
        try(
            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        ){
            StringTokenizer st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken());
            int L = Integer.parseInt(st.nextToken());
            int W = Integer.parseInt(st.nextToken());
            int H = Integer.parseInt(st.nextToken());

            System.out.println(solve(N, L, W, H));
        }
        catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: baekjoon/1166/1166.py
import sys


def solve(N, L, W, H):
    start, end = 0, max(L, W, H)

    for i in range(10000):
        mid = (start + end) / 2

        if (L//mid) * (W//mid) * (H//mid) >= N:
            start = mid
        else:
            end = mid

    return end


def main():
    N, L, W, H = map(int, sys.stdin.readline().split())
    answer = solve(N, L, W, H)
    print("%.10f" % answer)


if __name__ == "__main__":
    main()
